javalib/src/main/scala/java/util/TreeSet.scala

```diff
@@ -0,0 +1,222 @@
+package java.util
+
+import java.lang.Comparable
+import scala.collection.mutable
+import scala.math.Ordering
+
+import scala.collection.JavaConversions._
+
+class TreeSet[E] protected (ordering: Ordering[_ >: E], _comparator: Comparator[_ >: E])
+    extends AbstractSet[E]
+    with NavigableSet[E]
+    with Cloneable
+    with Serializable { self =>
+
+  def this() =
+    this(defaultOrdering[E], null.asInstanceOf[Comparator[_ >: E]])
+
+  def this(comparator: Comparator[_ >: E]) =
+    this(Ordering.comparatorToOrdering(comparator), null.asInstanceOf[Comparator[E]])
+
+  def this(collection: Collection[_ <: E]) = {
+    this(defaultOrdering[E], null.asInstanceOf[Comparator[E]])
+    addAll(collection)
+  }
+
+  def this(sortedSet: SortedSet[E]) = {
+    this(Ordering.comparatorToOrdering(sortedSet.comparator()),
+        sortedSet.comparator())
+    addAll(sortedSet)
+  }
+
+  private implicit object BoxOrdering extends Ordering[Box[E]] {
+    def compare(a: Box[E], b:Box[E]): Int = ordering.compare(a.inner, b.inner)
+  }
+
+  private val inner: mutable.TreeSet[Box[E]] = new mutable.TreeSet[Box[E]]()
+
+  def iterator(): Iterator[E] = {
+    new Iterator[E] {
+      private val iter = inner.clone.iterator
+
+      private var last: Option[E] = None
+
+      def hasNext(): Boolean = iter.hasNext
+
+      def next(): E = {
+        last = Some(iter.next().inner)
+        last.get
+      }
+
+      def remove(): Unit = {
+        if (last.isEmpty) {
+          throw new IllegalStateException()
+        } else {
+          last.foreach(self.remove(_))
+          last = None
+        }
+      }
+    }
+  }
+
+  def descendingIterator(): Iterator[E] = {
+    new Iterator[E] {
+      private val iter = inner.iterator.toList.reverse.iterator
+
+      private var last: Option[E] = None
+
+      def hasNext(): Boolean = iter.hasNext
+
+      def next(): E = {
+        last = Some(iter.next().inner)
+        last.get
+      }
+
+      def remove(): Unit = {
+        if (last.isEmpty) {
+          throw new IllegalStateException()
+        } else {
+          last.foreach(self.remove(_))
+          last = None
+        }
+      }
+    }
+  }
+
+  def descendingSet(): NavigableSet[E] = {
+    val descSetFun = { () =>
+      val retSet = new mutable.TreeSet[Box[E]]()(BoxOrdering.reverse)
+      retSet.addAll(inner)
+      retSet
+    }
+    new NavigableView(this, descSetFun)
+  }
+
+  def size(): Int =
+    inner.size
+
+  override def isEmpty(): Boolean =
```

Comment:
In this case I want to avoid to call ```size``` which can be expensive operation on tree, instead the check of direct leaf of the root node could be cheaper